File: internal/nativecapabilities/runtime.mjs
// Pi remains responsible for the model loop, MCP transport and authentication.
import { readFileSync, writeFileSync, renameSync } from 'node:fs'
import { createRequire } from 'node:module'
import { join, resolve } from 'node:path'
import { createHash } from 'node:crypto'

function loadProjectMcpConfig(api, path, cwd) {
  if (!path) return { mcpServers: {} }
  const previous = process.env.PI_MCP_CONFIG_MODE
  const previousWarn = console.warn
  const failurePrefix = `Failed to load MCP config from ${resolve(path)}:`
  let failed = false
  process.env.PI_MCP_CONFIG_MODE = 'exclusive'
  console.warn = (...args) => { if (args[0] === failurePrefix) failed = true; previousWarn(...args) }
  let config
  try { config = api.loadMcpConfig(path, cwd) }
  finally {
    console.warn = previousWarn
    if (previous === undefined) delete process.env.PI_MCP_CONFIG_MODE
    else process.env.PI_MCP_CONFIG_MODE = previous
  }
  if (failed) throw new Error('Project MCP configuration is invalid')
  return config
}

function mergeMcpConfigs(inherited, project) {
  const imports = [...new Set([...(inherited.imports ?? []), ...(project.imports ?? [])])]
  return {
    mcpServers: { ...inherited.mcpServers, ...project.mcpServers },
    ...(imports.length ? { imports } : {}),
    ...(inherited.settings || project.settings ? { settings: { ...inherited.settings, ...project.settings } } : {}),
    ...(project.claudePlugins !== undefined ? { claudePlugins: project.claudePlugins } : inherited.claudePlugins !== undefined ? { claudePlugins: inherited.claudePlugins } : {}),
  }
}

export default async function capabilities(pi) {
  const manifestPath = process.env.CHORA_NATIVE_CAPABILITIES_MANIFEST
  if (!manifestPath) return
  const manifest = JSON.parse(readFileSync(manifestPath, 'utf8'))
  const state = { version: 1, projectID: manifest.projectID, configVersion: manifest.config.version, attemptID: process.env.CHORA_ATTEMPT_ID, observedAt: '', skills: manifest.inventory.skills, servers: manifest.inventory.servers, bridge: manifest.inventory.bridge, running: false }
  const publish = () => {
    state.observedAt = new Date().toISOString()
    const output = process.env.CHORA_NATIVE_CAPABILITIES_STATUS
    writeFileSync(output + '.tmp', JSON.stringify(state), { mode: 0o600 })
    renameSync(output + '.tmp', output)
  }
  const warn = (ctx, message) => { ctx.ui.notify(message, 'warning') }
  const observeSkills = () => {
    const commands = pi.getCommands().filter(item => item.source === 'skill')
    const loaded = new Set(commands.map(item => item.sourceInfo?.path))
    state.skills = state.skills.map(item => ({ ...item, status: !item.enabled ? 'disabled' : item.status === 'unavailable' ? 'unavailable' : loaded.has(item.path) ? 'available' : 'discovered' }))
  }
  pi.events.on('pi-mcp-adapter/status/v1', snapshot => {
    if (snapshot?.version !== 1 || !Array.isArray(snapshot.servers)) return
    // The bridge publishes an empty handoff/shutdown snapshot as well. Retain
    // last-seen per-server facts and let session lifecycle mark freshness.
    if (!snapshot.servers.length && state.servers.length) return
    const previous = new Map(state.servers.map(item => [item.name, item]))
    state.servers = snapshot.servers.map(item => ({ name: item.name, source: previous.get(item.name)?.source ?? 'user', enabled: !item.disabled, status: item.status, toolCount: item.status === 'connected' ? item.toolCount : 0 }))
    publish()
  })
  if (manifest.inventory.bridge.status === 'discovered') {
    try {
      const require = createRequire(join(manifest.piRoot, 'package.json'))
      const { createJiti } = require('jiti')
      const jiti = createJiti(import.meta.url, { interopDefault: false, alias: { '@earendil-works/pi-coding-agent': join(manifest.piRoot, 'dist/index.js') } })
      const { createMcpAdapter } = await jiti.import(join(manifest.inventory.bridge.path, 'index.ts'))
      const api = await import(join(manifest.inventory.bridge.path, 'dist/config.js'))
      const inheritedConfig = api.loadMcpConfig(undefined, manifest.cwd)
      const projectConfig = loadProjectMcpConfig(api, manifest.config.mcpConfigPath, manifest.cwd)
      const config = mergeMcpConfigs(inheritedConfig, projectConfig)
      if (createHash('sha256').update(JSON.stringify(config)).digest('hex') !== manifest.inventory.configurationDigest) throw new Error('MCP configuration changed after preparation')
      config.settings = { ...config.settings, autoAuth: false, scriptMode: false, sampling: false }
      for (const name of manifest.config.disabledMcpServers ?? []) if (config.mcpServers[name]) config.mcpServers[name].disabled = true
      createMcpAdapter({ config })(pi)
    } catch {
      state.bridge.status = 'unavailable'
      state.servers = state.servers.map(item => ({ ...item, status: item.enabled ? 'failed' : 'disabled', toolCount: 0 }))
    }
  }
  pi.on('session_start', (_event, ctx) => {
    state.running = true
    observeSkills()
    if (state.bridge.status === 'unavailable') warn(ctx, 'MCP bridge could not load. Its capabilities are unavailable; this task can continue with other tools.')
    publish()
  })
  pi.on('before_agent_start', (_event, ctx) => {
    observeSkills()
    const unavailable = [...state.skills, ...state.servers].filter(item => ['failed', 'needs-auth', 'unavailable'].includes(item.status))
    if (unavailable.length) warn(ctx, `${unavailable.length} configured capabilities are unavailable. Inspect Skills and MCP for this execution.`)
    publish()
  })
  pi.on('session_shutdown', () => { state.running = false; publish() })
}

File: internal/nativecapabilities/runtime.go
package nativecapabilities

import (
	"bytes"
	"context"
	"crypto/sha256"
	_ "embed"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"reflect"
	"strings"
	"time"

	"github.com/Yangyang96/chora/internal/domain"
	"github.com/Yangyang96/chora/internal/execution"
)

//go:embed discovery.mjs
var discoverySource []byte

//go:embed runtime.mjs
var runtimeSource []byte

type Capability struct {
	Name      string `json:"name"`
	Path      string `json:"path,omitempty"`
	Source    string `json:"source"`
	Enabled   bool   `json:"enabled"`
	Status    string `json:"status"`
	ToolCount int    `json:"toolCount,omitempty"`
}
type Bridge struct {
	Status  string `json:"status"`
	Path    string `json:"path"`
	Version string `json:"version"`
}
type Inventory struct {
	ConfigurationDigest string       `json:"configurationDigest,omitempty"`
	ResourceDigest      string       `json:"resourceDigest,omitempty"`
	Skills              []Capability `json:"skills"`
	Servers             []Capability `json:"servers"`
	Bridge              Bridge       `json:"bridge"`
	MissingPackages     int          `json:"missingPackages"`
	Diagnostics         []struct {
		Path   string `json:"path"`
		Status string `json:"status"`
	} `json:"diagnostics"`
	ExtensionPaths []string `json:"extensionPaths,omitempty"`
}
type Manifest struct {
	ProjectID string    `json:"projectID"`
	Config    Config    `json:"config"`
	PiRoot    string    `json:"piRoot"`
	Cwd       string    `json:"cwd"`
	AgentDir  string    `json:"agentDir"`
	Inventory Inventory `json:"inventory"`
	Arguments []string  `json:"arguments"`
}

func packageRoot(executable string) (string, error) {
	resolved, err := filepath.EvalSymlinks(executable)
	if err != nil {
		return "", err
	}
	for root, depth := filepath.Dir(resolved), 0; depth < 6; root, depth = filepath.Dir(root), depth+1 {
		var metadata struct {
			Name    string
			Version string
		}
		raw, e := os.ReadFile(filepath.Join(root, "package.json"))
		if e == nil && json.Unmarshal(raw, &metadata) == nil && metadata.Name == "@earendil-works/pi-coding-agent" && metadata.Version == "0.85.1" {
			return root, nil
		}
	}
	return "", errors.New("capability management requires the qualified Pi 0.85.1 package")
}

// Inspect does not import extension code or connect to MCP servers.
func Inspect(ctx context.Context, executable, agentDir, cwd string, config Config) (Inventory, error) {
	root, err := packageRoot(executable)
	if err != nil {
		return Inventory{}, err
	}
	input := Manifest{Config: config, PiRoot: root, Cwd: cwd, AgentDir: agentDir}
	raw, err := json.Marshal(input)
	if err != nil {
		return Inventory{}, err
	}
	ctx, cancel := context.WithTimeout(ctx, 15*time.Second)
	defer cancel()
	command := exec.CommandContext(ctx, "node", "--input-type=module", "-e", string(discoverySource), "--discover")
	// -e consumes the script without a filename, so argv[1] is the first argument.
	command.Args = append(command.Args[:len(command.Args)-1], "chora-discovery", "--discover")
	command.Stdin = bytes.NewReader(raw)
	command.Dir = cwd
	command.Env = append(os.Environ(), "PI_OFFLINE=1", "PI_CODING_AGENT_DIR="+agentDir)
	var output limitedOutput
	command.Stdout = &output
	if err = command.Run(); err != nil {
		return Inventory{}, errors.New("native capability discovery failed; inspect the selected Pi installation and configuration")
	}
	var inventory Inventory
	if err = json.Unmarshal(output.Bytes(), &inventory); err != nil {
		return Inventory{}, errors.New("invalid native capability discovery response")
	}
	return inventory, nil
}

type limitedOutput struct{ buffer bytes.Buffer }

func (output *limitedOutput) Bytes() []byte { return output.buffer.Bytes() }

func (output *limitedOutput) Write(data []byte) (int, error) {
	if output.buffer.Len()+len(data) > 1<<20 {
		return 0, errors.New("capability output exceeds limit")
	}
	return output.buffer.Write(data)
}

// Prepare creates a private immutable configuration reference for an execution.
// An explicit resume reuses its session owner's original Project configuration.
func Prepare(ctx context.Context, root, projectID, executable, agentDir, cwd string, attemptID, ownerID domain.AttemptID, config Config) ([]string, map[string]string, error) {
	if config.Version == 0 && !ownerID.Valid() {
		return nil, nil, nil
	}
	if _, err := domain.ParseProjectID(projectID); err != nil {
		return nil, nil, err
	}
	if !attemptID.Valid() {
		return nil, nil, errors.New("invalid capability attempt")
	}
	if err := os.MkdirAll(root, 0700); err != nil {
		return nil, nil, err
	}
	canonical, err := filepath.EvalSymlinks(root)
	if err != nil || canonical != root {
		return nil, nil, errors.New("capability directory is not canonical")
	}
	sourceHash := sha256.Sum256(runtimeSource)
	extension := filepath.Join(root, fmt.Sprintf("runtime-%x.mjs", sourceHash))
	if err = writeImmutable(extension, runtimeSource); err != nil {
		return nil, nil, err
	}
	identity := attemptID
	if ownerID.Valid() {
		identity = ownerID
	}
	path := filepath.Join(root, identity.String()+".json")
	var manifest Manifest
	if ownerID.Valid() {
		raw, e := readPrivate(path)
		if os.IsNotExist(e) && config.Version == 0 {
			return nil, nil, nil
		}
		if e != nil {
			return nil, nil, errors.New("original capability configuration unavailable for resume")
		}
		if json.Unmarshal(raw, &manifest) != nil || manifest.ProjectID != projectID || manifest.Cwd != cwd {
			return nil, nil, errors.New("resume capability configuration mismatch")
		}
		current, e := Inspect(ctx, executable, agentDir, cwd, manifest.Config)
		if e != nil || current.ResourceDigest != manifest.Inventory.ResourceDigest {
			return nil, nil, errors.New("native capability resources changed; start a fresh execution instead of resuming this session")
		}
	} else {
		piRoot, e := packageRoot(executable)
		if e != nil {
			return nil, nil, e
		}
		inventory, e := Inspect(ctx, executable, agentDir, cwd, config)
		if e != nil {
			return nil, nil, e
		}
		manifest = Manifest{ProjectID: projectID, Config: config, PiRoot: piRoot, AgentDir: agentDir, Cwd: cwd, Inventory: inventory}
		manifest.Arguments = []string{"--offline", "--no-skills", "--no-extensions"}
		for _, skill := range inventory.Skills {
			if skill.Enabled && skill.Status != "unavailable" {
				manifest.Arguments = append(manifest.Arguments, "--skill", skill.Path)
			}
		}
		for _, extensionPath := range inventory.ExtensionPaths {
			manifest.Arguments = append(manifest.Arguments, "--extension", extensionPath)
		}
		manifest.Arguments = append(manifest.Arguments, "--extension", extension)
		raw, e := json.Marshal(manifest)
		if e != nil {
			return nil, nil, e
		}
		if e = writeImmutable(path, raw); e != nil {
			return nil, nil, e
		}
	}
	raw, err := readPrivate(path)
	if err != nil {
		return nil, nil, err
	}
	sum := sha256.Sum256(raw)
	environment := map[string]string{"CHORA_NATIVE_CAPABILITIES_MANIFEST": path, "CHORA_NATIVE_CAPABILITIES_SHA256": hex.EncodeToString(sum[:]), "CHORA_NATIVE_CAPABILITIES_STATUS": filepath.Join(root, attemptID.String()+".status.json")}
	return manifest.Arguments, environment, nil
}
func writeImmutable(path string, body []byte) error {
	file, err := os.OpenFile(path, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0600)
	if errors.Is(err, os.ErrExist) {
		previous, e := readPrivate(path)
		if e != nil {
			return e
		}
		if !bytes.Equal(previous, body) {
			return errors.New("capability configuration changed")
		}
		return nil
	}
	if err != nil {
		return err
	}
	_, err = file.Write(body)
	return errors.Join(err, file.Close())
}
func readPrivate(path string) ([]byte, error) {
	info, err := os.Lstat(path)
	if err != nil {
		return nil, err
	}
	if !info.Mode().IsRegular() || info.Mode().Perm()&0077 != 0 || info.Size() > 1<<20 {
		return nil, errors.New("unsafe capability file")
	}
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()
	return io.ReadAll(io.LimitReader(file, (1<<20)+1))
}

// StripValidatedArguments permits only the exact suffix prepared for this child.
// The supervisor still validates the original native Pi argv and observer.
func StripValidatedArguments(root string, invocation execution.Invocation) ([]string, error) {
	env := invocation.Environment()
	path := env["CHORA_NATIVE_CAPABILITIES_MANIFEST"]
	if path == "" {
		return invocation.Arguments(), nil
	}
	if filepath.Dir(path) != root || !strings.HasSuffix(path, ".json") {
		return nil, errors.New("capability manifest escaped runtime root")
	}
	raw, err := readPrivate(path)
	if err != nil {
		return nil, err
	}
	sum := sha256.Sum256(raw)
	if hex.EncodeToString(sum[:]) != env["CHORA_NATIVE_CAPABILITIES_SHA256"] {
		return nil, errors.New("capability manifest identity changed")
	}
	var manifest Manifest
	if json.Unmarshal(raw, &manifest) != nil || manifest.Cwd != invocation.WorkingRoot() {
		return nil, errors.New("capability manifest context mismatch")
	}
	sourceHash := sha256.Sum256(runtimeSource)
	expected := filepath.Join(root, fmt.Sprintf("runtime-%x.mjs", sourceHash))
	source, err := readPrivate(expected)
	if err != nil || !bytes.Equal(source, runtimeSource) {
		return nil, errors.New("capability observer identity changed")
	}
	args := invocation.Arguments()
	count := len(manifest.Arguments)
	if count == 0 || len(args) < count || !reflect.DeepEqual(args[len(args)-count:], manifest.Arguments) {
		return nil, errors.New("capability arguments differ from prepared configuration")
	}
	status := env["CHORA_NATIVE_CAPABILITIES_STATUS"]
	if filepath.Dir(status) != root || filepath.Base(status) != env["CHORA_ATTEMPT_ID"]+".status.json" {
		return nil, errors.New("capability status identity mismatch")
	}
	return args[:len(args)-count], nil
}
